ledger/accountdb_test.go

```diff
@@ -2131,6 +2131,179 @@ func TestResourcesDataSetData(t *testing.T) {
 	}
 }
 
+// TestResourceDataRoundtripConversion ensures that basics.AppLocalState, basics.AppParams,
+// basics.AssetHolding, and basics.AssetParams can be converted to resourcesData and back without
+// losing any data. It uses reflection to be sure that no new fields are omitted.
+//
+// In other words, this test makes sure any new fields in basics.AppLocalState, basics.AppParams,
+// basics.AssetHolding, or basics.AssetParam also get added to resourcesData.
+func TestResourceDataRoundtripConversion(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	t.Run("basics.AppLocalState", func(t *testing.T) {
+		t.Parallel()
+		for i := 0; i < 1000; i++ {
+			randObj, _ := protocol.RandomizeObject(&basics.AppLocalState{})
+			basicsAppLocalState := *randObj.(*basics.AppLocalState)
+
+			var data resourcesData
+			data.SetAppLocalState(basicsAppLocalState)
+			roundTripAppLocalState := data.GetAppLocalState()
+
+			require.Equal(t, basicsAppLocalState, roundTripAppLocalState)
+		}
+	})
+
+	t.Run("basics.AppParams", func(t *testing.T) {
+		t.Parallel()
+		for i := 0; i < 1000; i++ {
+			randObj, _ := protocol.RandomizeObject(&basics.AppParams{})
+			basicsAppParams := *randObj.(*basics.AppParams)
+
+			var data resourcesData
+			data.SetAppParams(basicsAppParams, false)
```

Comment:
@jasonpaulos Sanity checking:  Is there a rationale for _not_ testing `haveHoldings = true` like `data.SetAppParams(basicsAppParams, true)`?

Reply:
@jasonpaulos It's possible someone changes the implementation in the future, so I think it's prudent to support generating all possible inputs.  

I'm good with either a 2nd hard-coded invocation or extending the random generation strategy.

Note - The concern applies to `SetAssetParams`.

Reply:
The concern _also_ applies to*

Reply:
Added in d6f6199fd